### PR TITLE
refactor: migrate PDFView/index.js to function component

### PR DESCRIPTION
**Pull request by dummy-1111**
### Details

Migrate PDFView/index.js to function component

### Fixed Issues

$ https://github.com/Expensify/App/issues/16186
PROPOSAL: https://github.com/Expensify/App/issues/16186#issuecomment-1578490998

### Tests
1. Open ND and log in with any account
2. Navigate to any chat
3. [WEB AND MOBILE WEB] Tap "+" button near the main composer and select "Add attachment"
4. [WEB AND MOBILE WEB] Attach normal pdf document and verify that the pdf is displayed properly in preview
5. [WEB AND MOBILE WEB] Send the pdf
6. [WEB AND MOBILE WEB] Attach protected pdf document and verify that the pdf is displayed properly in preview
7. [WEB AND MOBILE WEB] Send the pdf
8. [WEB AND MOBILE WEB] Tap pdf documents in the chat and they are displayed properly

- [x] Verify that no errors appear in the JS console

### Offline tests
1. Open ND and log in with any account
2. Navigate to any chat
3. [WEB AND MOBILE WEB] Tap "+" button near the main composer and select "Add attachment"
4. [WEB AND MOBILE WEB] Attach normal pdf document and verify that the pdf is displayed properly in preview
5. [WEB AND MOBILE WEB] Send the pdf
6. [WEB AND MOBILE WEB] Attach protected pdf document and verify that the pdf is displayed properly in preview
7. [WEB AND MOBILE WEB] Send the pdf
8. [WEB AND MOBILE WEB] Tap pdf documents in the chat and they are displayed properly

### QA Steps
1. Open ND and log in with any account
2. Navigate to any chat
3. [WEB AND MOBILE WEB] Tap "+" button near the main composer and select "Add attachment"
4. [WEB AND MOBILE WEB] Attach normal pdf document and verify that the pdf is displayed properly in preview
5. [WEB AND MOBILE WEB] Send the pdf
6. [WEB AND MOBILE WEB] Attach protected pdf document and verify that the pdf is displayed properly in preview
7. [WEB AND MOBILE WEB] Send the pdf
8. [WEB AND MOBILE WEB] Tap pdf documents in the chat and they are displayed properly

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/126839717/5fcc2ef0-828b-432b-9cc4-09aef7d7b4ea

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/126839717/13681f0a-ea29-4620-b37f-773a74c5af31

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/126839717/f48b1584-0585-4f65-af60-2704196828f2

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/126839717/dcba0292-8c21-4173-9dcf-55c4005a6e4d

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/126839717/b54a0624-20b8-441d-b799-11cdf3b47109

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/126839717/9b5afcd0-04fb-44cd-9edd-a4c14ec6cdcf

</details>

### CODE REVIEW

**Comment by s-alves10** [comment]
@ishpaul777 

friendly bump

File: src/components/PDFView/index.js
Comment (s-alves10) [reply src/components/PDFView/index.js:188]: Actually we should have `pageViewports` as its dependency.
But `pageViewports` is an object(array) and changed every time the document is loaded. Having it as a state and specifying as the dependency of this method would re-render the component every time the document is loaded and would cause the component infinite re-render. And having this as a ref wouldn't change the `calculatePageWidth` method after the document is loaded.

As you can see, `pageViewports` and `numPages` are set when the document is loaded, but `numPages` is a scalar value and not changed as long as the document is not changed. So specifying this as a dependency instead of `pageViewports` would prevent the infinite re-render. And when the document is changed, it would change as well

This is the reason I added `numPages` as a dependency instead of `pageViewports`

**Comment by s-alves10** [comment]
@ishpaul777 friendly bump

File: src/components/PDFView/index.js
Comment (s-alves10) [reply src/components/PDFView/index.js:188]: done

File: src/components/PDFView/index.js
Comment (s-alves10) [reply src/components/PDFView/index.js]: done

**Comment by s-alves10** [comment]
Please take a look again @ishpaul777

**Comment by s-alves10** [comment]
Let me check @ishpaul777

**Comment by s-alves10** [comment]
@ishpaul777 

> I found while adjusting window width, the pages has a lot of space between them can you give a quick look to see if this comes from this PR (becuase i can't repro on staging)

I can reproduce this issue on staging

https://github.com/Expensify/App/assets/126839717/a22adc1c-258d-4f0a-afa7-f7428010b8ee

**Comment by s-alves10** [comment]
@ishpaul777 

I've fixed the above errors related to password. Please take a look

**Comment by s-alves10** [comment]
Can you please take a look again? @ishpaul777

**Review by ishpaul777 — APPROVED** [review]
just add the missing comment back rest LGTM 👍

File: src/components/PDFView/index.js
Comment (puneetlath) [reply src/components/PDFView/index.js:39]: Why?

File: src/components/PDFView/index.js
Comment (puneetlath) [inline src/components/PDFView/index.js:138]: Could we move this function outside? So we don't have to memoize it? And just pass it the props it needs to do the calculations as arguments?

File: src/components/PDFView/index.js
Comment (puneetlath) [inline src/components/PDFView/index.js:156]: Same question here. Can we move it outside the component function?

File: src/components/PDFView/index.js
Comment (puneetlath) [inline src/components/PDFView/index.js:171]: Same question about this function.

**Review by puneetlath — CHANGES_REQUESTED** [review]

**Comment by puneetlath** [comment]
@ishpaul777 can you re-review and then ping me again when you're ready for me to take another look?

**Comment by puneetlath** [comment]
@s-alves10 will you be able to take this to the finish line? If we don't hear back by tomorrow I'm going to proceed with hiring someone else.

**Comment by s-alves10** [comment]
@puneetlath 
Due to my health, I am not able to continue working for a while. Please go with other one.

**Comment by puneetlath** [comment]
Sorry to hear that @s-alves10 😞

**Comment by puneetlath** [comment]
Good call. Closing it out and putting the issue on hold.

**Pull request closed without merging**
